File: src/PicLens.Application/Services/FileRenamePlanner.cs
using PicLens.Core.Domain;

namespace PicLens.Application.Services;

public static class FileRenamePlanner
{
    private const string AlreadyTargetSequenceReason = "already_target_sequence";

    public static FileNameValidationResult ValidateImageFileName(string fileName)
    {
        if (string.IsNullOrWhiteSpace(fileName))
        {
            return new FileNameValidationResult(false, "empty_name");
        }

        if (fileName.IndexOfAny(Path.GetInvalidFileNameChars()) >= 0 || fileName.Contains('/') || fileName.Contains('\\'))
        {
            return new FileNameValidationResult(false, "invalid_file_name");
        }

        return ImageFormatRules.GetSupportedImageExtension(fileName) is null
            ? new FileNameValidationResult(false, "unsupported_extension")
            : new FileNameValidationResult(true, null);
    }

    public static DropTargetBatchRenamePlan PlanDropTargetBatchRename(
        IEnumerable<string> sourcePaths,
        string targetPath,
        Func<string, string, bool> targetNameExists)
    {
        ArgumentNullException.ThrowIfNull(sourcePaths);
        ArgumentNullException.ThrowIfNull(targetNameExists);

        var targetDirectory = Path.GetDirectoryName(targetPath)
            ?? throw new ArgumentException("Target path must include a directory.", nameof(targetPath));
        var targetBaseName = Path.GetFileNameWithoutExtension(targetPath);

        var items = new List<DropTargetBatchRenamePlanItem>();
        var sequenceNumber = 1;

        foreach (var source in sourcePaths.Where(source => !PathEquals(source, targetPath)))
        {
            var item = CreatePlanItem(source, targetDirectory, targetBaseName, sequenceNumber, targetNameExists);
            items.Add(item);
            sequenceNumber = Math.Max(sequenceNumber, ExtractSequenceNumber(item.TargetPath, targetBaseName) + 1);
        }

        return new DropTargetBatchRenamePlan(items.Count, items);
    }

    private static DropTargetBatchRenamePlanItem CreatePlanItem(
        string sourcePath,
        string targetDirectory,
        string targetBaseName,
        int sequenceNumber,
        Func<string, string, bool> targetNameExists)
    {
        var sourceSequenceNumber = TryExtractSequenceNumber(sourcePath, targetBaseName);
        if (sourceSequenceNumber is not null
            && sourceSequenceNumber.Value < sequenceNumber
            && !targetNameExists(sourcePath, sourcePath))
        {
            return new DropTargetBatchRenamePlanItem(sourcePath, sourcePath, true, AlreadyTargetSequenceReason);
        }

        var nextTargetPath = NextAvailableSequenceTargetPath(
            sourcePath,
            targetDirectory,
            targetBaseName,
            sequenceNumber,
            targetNameExists);
        return PathEquals(sourcePath, nextTargetPath)
            ? new DropTargetBatchRenamePlanItem(sourcePath, nextTargetPath, true, AlreadyTargetSequenceReason)
            : new DropTargetBatchRenamePlanItem(sourcePath, nextTargetPath, false, null);
    }

    private static string NextAvailableSequenceTargetPath(
        string sourcePath,
        string targetDirectory,
        string targetBaseName,
        int sequenceNumber,
        Func<string, string, bool> targetNameExists)
    {
        var candidateSequence = sequenceNumber;
        while (true)
        {
            var candidatePath = CreateSequenceTargetPath(sourcePath, targetDirectory, targetBaseName, candidateSequence);
            if (!targetNameExists(candidatePath, sourcePath))
            {
                return candidatePath;
            }

            candidateSequence += 1;
        }
    }

    private static string CreateSequenceTargetPath(
        string sourcePath,
        string targetDirectory,
        string targetBaseName,
        int sequenceNumber) =>
        Path.Combine(targetDirectory, $"{targetBaseName}-{sequenceNumber:00}{Path.GetExtension(sourcePath)}");

    private static int ExtractSequenceNumber(string targetPath, string targetBaseName)
    {
        return TryExtractSequenceNumber(targetPath, targetBaseName)
            ?? throw new ArgumentException("Target path must include a target sequence number.", nameof(targetPath));
    }

    private static int? TryExtractSequenceNumber(string targetPath, string targetBaseName)
    {
        var targetName = Path.GetFileNameWithoutExtension(targetPath);
        if (!targetName.StartsWith($"{targetBaseName}-", StringComparison.OrdinalIgnoreCase))
        {
            return null;
        }

        var suffix = targetName[(targetBaseName.Length + 1)..];
        return suffix.Length > 0 && suffix.All(char.IsDigit) && int.TryParse(suffix, out var sequenceNumber)
            ? sequenceNumber
            : null;
    }

    private static bool PathEquals(string left, string right) =>
        string.Equals(
            Path.GetFullPath(left),
            Path.GetFullPath(right),
            OperatingSystem.IsWindows() ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal);

}

public sealed record FileNameValidationResult(bool IsValid, string? Reason);

public sealed record DropTargetBatchRenamePlan(int Total, IReadOnlyList<DropTargetBatchRenamePlanItem> Items);

public sealed record DropTargetBatchRenamePlanItem(
    string SourcePath,
    string TargetPath,
    bool ShouldSkip,
    string? Reason);

File: qt/src/core/src/file_rename_planner.cpp
#include <piclens/core/file_rename_planner.h>

#include <piclens/core/image_format_rules.h>
#include <piclens/core/path_rules.h>

#include <QDir>
#include <QFileInfo>
#include <QSet>

#include <algorithm>
#include <stdexcept>

namespace piclens::core::file_rename_planner {
namespace {

bool containsReservedFileNameCharacter(const QString &fileName)
{
    static const QSet<QChar> reservedCharacters{
        QLatin1Char('<'),
        QLatin1Char('>'),
        QLatin1Char(':'),
        QLatin1Char('"'),
        QLatin1Char('/'),
        QLatin1Char('\\'),
        QLatin1Char('|'),
        QLatin1Char('?'),
        QLatin1Char('*'),
    };

    return std::any_of(fileName.cbegin(), fileName.cend(), [&](QChar character) {
        return reservedCharacters.contains(character)
            || character.category() == QChar::Other_Control;
    });
}

std::optional<int> tryExtractSequenceNumber(
    const QString &targetPath,
    const QString &targetBaseName)
{
    const QString targetName = QFileInfo(targetPath).completeBaseName();
    const QString prefix = targetBaseName + QLatin1Char('-');
    if (!targetName.startsWith(prefix, path_rules::pathCaseSensitivity())) {
        return std::nullopt;
    }

    const QString suffix = targetName.sliced(prefix.size());
    if (suffix.isEmpty()
        || !std::all_of(suffix.cbegin(), suffix.cend(), [](QChar character) {
               return character.isDigit();
           })) {
        return std::nullopt;
    }

    bool parsed = false;
    const int sequenceNumber = suffix.toInt(&parsed);
    return parsed ? std::optional<int>{sequenceNumber} : std::nullopt;
}

int extractSequenceNumber(const QString &targetPath, const QString &targetBaseName)
{
    const auto value = tryExtractSequenceNumber(targetPath, targetBaseName);
    if (!value.has_value()) {
        throw std::invalid_argument("Target path must include a target sequence number.");
    }
    return *value;
}

QString createSequenceTargetPath(
    const QString &sourcePath,
    const QString &targetDirectory,
    const QString &targetBaseName,
    int sequenceNumber)
{
    const QString sourceSuffix = QFileInfo(sourcePath).suffix();
    const QString extension = sourceSuffix.isEmpty()
        ? QString()
        : QLatin1Char('.') + sourceSuffix;
    const QString fileName = QStringLiteral("%1-%2%3")
                                 .arg(targetBaseName)
                                 .arg(sequenceNumber, 2, 10, QLatin1Char('0'))
                                 .arg(extension);
    return QDir(targetDirectory).filePath(fileName);
}

QString nextAvailableSequenceTargetPath(
    const QString &sourcePath,
    const QString &targetDirectory,
    const QString &targetBaseName,
    int sequenceNumber,
    const QVector<QString> &existingPaths)
{
    int candidateSequence = sequenceNumber;
    while (true) {
        const QString candidatePath = createSequenceTargetPath(
            sourcePath,
            targetDirectory,
            targetBaseName,
            candidateSequence);
        if (!path_rules::targetNameExists(existingPaths, candidatePath, sourcePath)) {
            return candidatePath;
        }
        ++candidateSequence;
    }
}

DropTargetBatchRenamePlanItem createPlanItem(
    const QString &sourcePath,
    const QString &targetDirectory,
    const QString &targetBaseName,
    int sequenceNumber,
    const QVector<QString> &existingPaths)
{
    const auto sourceSequenceNumber = tryExtractSequenceNumber(sourcePath, targetBaseName);
    if (sourceSequenceNumber.has_value()
        && *sourceSequenceNumber < sequenceNumber
        && !path_rules::targetNameExists(existingPaths, sourcePath, sourcePath)) {
        return {
            .sourcePath = sourcePath,
            .targetPath = sourcePath,
            .shouldSkip = true,
            .reason = AlreadyTargetSequenceReason,
        };
    }

    const QString nextTargetPath = nextAvailableSequenceTargetPath(
        sourcePath,
        targetDirectory,
        targetBaseName,
        sequenceNumber,
        existingPaths);
    if (path_rules::pathEquals(sourcePath, nextTargetPath)) {
        return {
            .sourcePath = sourcePath,
            .targetPath = nextTargetPath,
            .shouldSkip = true,
            .reason = AlreadyTargetSequenceReason,
        };
    }

    return {
        .sourcePath = sourcePath,
        .targetPath = nextTargetPath,
        .shouldSkip = false,
        .reason = std::nullopt,
    };
}

} // namespace

FileNameValidationResult validateImageFileName(const QString &fileName)
{
    if (fileName.trimmed().isEmpty()) {
        return {.reason = QStringLiteral("empty_name")};
    }

    if (containsReservedFileNameCharacter(fileName)) {
        return {.reason = QStringLiteral("invalid_file_name")};
    }

    if (!image_format_rules::supportedImageExtension(fileName).has_value()) {
        return {.reason = QStringLiteral("unsupported_extension")};
    }

    return {.isValid = true, .reason = std::nullopt};
}

DropTargetBatchRenamePlan planDropTargetBatchRename(
    const QVector<QString> &sourcePaths,
    const QString &targetPath,
    const QVector<QString> &existingPaths)
{
    const QFileInfo targetInfo(targetPath);
    const QString targetDirectory = targetInfo.absolutePath();
    const QString targetBaseName = targetInfo.completeBaseName();

    QVector<DropTargetBatchRenamePlanItem> items;
    int sequenceNumber = 1;

    for (const QString &sourcePath : sourcePaths) {
        if (path_rules::pathEquals(sourcePath, targetPath)) {
            continue;
        }

        DropTargetBatchRenamePlanItem item = createPlanItem(
            sourcePath,
            targetDirectory,
            targetBaseName,
            sequenceNumber,
            existingPaths);
        sequenceNumber = std::max(
            sequenceNumber,
            extractSequenceNumber(item.targetPath, targetBaseName) + 1);
        items.append(std::move(item));
    }

    return {
        .total = static_cast<int>(items.size()),
        .items = std::move(items),
    };
}

} // namespace piclens::core::file_rename_planner
